src/gui/TextEditor.cpp

```diff
@@ -960,53 +961,47 @@ void TextEditor::paint(cairo_t* cr, GdkRectangle* repaintRect, double zoom) {
             pos += gtk_text_iter_get_bytes_in_line(&cursorIter);
         }
         gtk_text_iter_set_offset(&cursorIter, offset);
-
         string txt = text.substr(0, pos) + preeditString + text.substr(pos);
 
-        PangoAttribute* attrib = pango_attr_underline_new(PANGO_UNDERLINE_SINGLE);
-        PangoAttrList* list = pango_layout_get_attributes(this->layout);
-
-        attrib->start_index = pos;
-        attrib->end_index = pos + preeditString.length();
-
-        if (list == nullptr) {
-            list = pango_attr_list_new();
-            pango_layout_set_attributes(this->layout, list);
+        PangoAttrList* list = pango_attr_list_new();
+        PangoAttrList* preedit_list = this->preeditAttrList;
+        if (list != nullptr) {
+            pango_attr_list_splice(list, preedit_list, pos, preeditString.length());
         }
-        pango_attr_list_insert(list, attrib);
-
+        pango_layout_set_attributes(this->layout, list);
         pango_layout_set_text(this->layout, txt.c_str(), txt.length());
     } else {
         string txt = this->text->getText();
         pango_layout_set_text(this->layout, txt.c_str(), txt.length());
-    }
 
-    GtkTextIter start;
-    GtkTextIter end;
-    bool hasSelection = gtk_text_buffer_get_selection_bounds(this->buffer, &start, &end);
+        GtkTextIter start;
+        GtkTextIter end;
+        bool hasSelection = gtk_text_buffer_get_selection_bounds(this->buffer, &start, &end);
 
-    if (hasSelection) {
-        auto selectionColorU16 = Util::GdkRGBA_to_ColorU16(selectionColor);
-        PangoAttribute* attrib =
-                pango_attr_background_new(selectionColorU16.red, selectionColorU16.green, selectionColorU16.blue);
-        PangoAttrList* list = pango_layout_get_attributes(this->layout);
+        if (hasSelection) {
+            auto selectionColorU16 = Util::GdkRGBA_to_ColorU16(selectionColor);
+            PangoAttribute* attrib =
+                    pango_attr_background_new(selectionColorU16.red, selectionColorU16.green, selectionColorU16.blue);
+            PangoAttrList* list = pango_layout_get_attributes(this->layout);
 
-        attrib->start_index = getByteOffset(gtk_text_iter_get_offset(&start));
-        attrib->end_index = getByteOffset(gtk_text_iter_get_offset(&end));
+            attrib->start_index = getByteOffset(gtk_text_iter_get_offset(&start));
+            attrib->end_index = getByteOffset(gtk_text_iter_get_offset(&end));
 
-        if (list == nullptr) {
-            list = pango_attr_list_new();
+            if (list == nullptr) {
+                list = pango_attr_list_new();
+                pango_layout_set_attributes(this->layout, list);
+            }
+            pango_attr_list_insert(list, attrib);
+        } else {
+            // remove all attributes
+            PangoAttrList* list = pango_attr_list_new();
             pango_layout_set_attributes(this->layout, list);
```

Comment:
As far as I understand, we need a `pango_attr_list_unref(list)` here since `list` is owned here and `pango_layout_set_attributes` only borrows `list`.

Reply:
I agree.